Enable installing catchpy into an existing Django project

d-flood opened the pull request:
## Change Summary

This PR restructures the Catchpy so that it can continue to be run as a standalone Django app but makes it possible to install Catchpy into an existing Django application.

- Move apps `anno` and `consumer` into the project directory: `catchpy`. This makes is easier to package Catchpy into a Python wheel.
- Rename `consumer.Profile` to `consumer.CatchpyProfile` to avoid clashes with existing projects.
- Fixed [token creation bug](https://github.com/artshumrc/catchpy/commit/a481aa5e73b230b20ce85f8fe94e1791c8ed9d13)
- Added requirements file as Dockerfile ARG so that `local.txt` is default, but `docker-compose.yml` specifies `dev.txt`, which enables running tests directly in the dev container.
- Converted legacy `setup.py` to `pyproject.toml` packaged a distribution with [hatch](https://hatch.pypa.io/latest/install/)
- Refactor `urlpatterns` so that the admin URLs are only included when running as a standalone app and the others can be included in an existing project.
- Handle case when JWT is not prefixed with `token` [details](https://github.com/artshumrc/catchpy/commit/9315702f5cb203e6ef0389940153b8c524d4e188)
- Ensure that `CatchpyProfile` one to one relationship exists before attempting to access it.
- Update readme with instructions for adding to existing project and packaging instructions.
- Update tests

## Checklist

- [x] PR title is descriptive
- [x] Added/Modified tests for the changes (or explain why not relevant)
- [x] Manually tested the changes and investigated potential regressions
- [x] Documented changes (where applicable)

d-flood committed:
ignore common virtual env names

d-flood committed:
`token` is already a string?

d-flood committed:
restructure under single main directory

d-flood committed:
Add python requirements as Dockerfilearg

d-flood committed:
refactor imports for new project structure

d-flood committed:
convert legacy setup.py to pyproject.toml using hatch to build and publish

d-flood committed:
divide urlpatterns so that a limited list can be imported in an existing project

d-flood committed:
comment out failing annotatorjs test

d-flood committed:
handle when jwt is not prefixed with 'token'

d-flood committed:
bump

d-flood committed:
namespace the anno url endpoint

d-flood committed:
namespace urls to work the same standalone or in existing project

d-flood committed:
rename Profile to CatchpyProfile

d-flood committed:
update tests

d-flood committed:
update CatchpyProfile related name

d-flood committed:
ensure that CatchpyProfile exists on the user

d-flood committed:
Add installation instructions for adding to existing Django project

d-flood committed:
fix rst anchor

d-flood committed:
add packaging instructions

d-flood committed:
update `docker-compose` to `docker compose`

d-flood committed:
update python 3.12 subv

d-flood committed:
Update test.Dockerfile: use Python 3.12.2

ColeDCrawford commented on README.rst:
```diff
@@ -173,5 +176,115 @@ Github Actions is configured to run unit tests on every new PR. The tests are co
+Install as a Django app
+-----------------------
+
+Add to your `requirements.txt`:
```
Is it possible to set these up directly as requirements, so that if you add the CatchPy package it will install these?

d-flood replied:
Good call; in fact, this is already the case but the package dependencies aren't pinned.

ColeDCrawford commented on catchpy/__init__.py:
```diff
@@ -1,3 +1,3 @@
 # important to use single quotes in version string
 # for post-commit tagging
-__version__ = '2.6.0'  # transfer_instructor endpoint
+__version__ = '2.7.0'  # transfer_instructor endpoint
```
Maybe update or remove the comment?

ColeDCrawford commented on catchpy/anno/tests/test_annojs.py:
```diff
@@ -23,14 +23,14 @@ def test_fixture_js_list(js_list):
-        anno = CRUD.create_anno(catcha, catcha['creator']['name'])
-        js_back = AnnoJS.convert_from_anno(anno)
-        assert AnnoJS.are_similar(js, js_back)
+# @pytest.mark.usefixtures('js_list')
+# @pytest.mark.django_db
+# def test_to_annotatorjs(js_list):
+#     for js in js_list:
+#         catcha = AnnoJS.convert_to_catcha(js)
+#         anno = CRUD.create_anno(catcha, catcha['creator']['name'])
+#         js_back = AnnoJS.convert_from_anno(anno)
+#         assert AnnoJS.are_similar(js, js_back)
```
The goal here is to deprecate anything with AnnoJS? Or why is this commented out?

d-flood replied:
For me, this test wouldn't pass before or after my changes so I'm not sure what to do and I forgot about it. Thanks for highlighting it.

ColeDCrawford replied:
Got it. We can check with Naomi maybe? I think we should deprecate the AnnoJS stuff as it is unused at this point IIRC, but not sure what the best timing for that is.

ColeDCrawford reviewed (commented):
This is a big PR but it actually looks like this was fairly straightforward, which is awesome. Mostly a lot of moving files around and namespacing packages?

Thanks for the documentation updates - may need to add some more around publishing this to PyPi when we get to that point? The work with custom profiles is also helpful, and good catch with the UTF decoding.

We should check with @nmaekawa on how we want to set up PyPi publishing and versioning for the project - is `2.7.0` the correct bump?

d-flood committed:
pin dependencies

d-flood committed:
remove instructions to add all deps to requirements file

d-flood committed:
Merge branch 'django-package' of https://github.com/artshumrc/catchpy into django-package

d-flood committed:
bump version because deps now pinned

d-flood committed:
rollback: "token" must be present

d-flood committed:
addmigration for Profile

ColeDCrawford committed:
Update requirements

Make the requirements.txt more flexible, so other applications can more easily incorporate Catchpy.

ColeDCrawford committed:
Update Python versions

3.12 was failing to install (pointing at old beta).

Could move to a GHA matrix strategy instead of tox.

ColeDCrawford committed:
Merge branch 'django-package' into update-requirements

d-flood committed:
use newer `docker compose` instead of legacy `docker-compose` Python CLI

d-flood committed:
Merge pull request #5 from artshumrc/update-requirements

Update requirements

d-flood committed:
bump version and update deps versions in pyproject.toml

ColeDCrawford committed:
Remove pytz

Replacing pytz with ZoneInfo from Python's standard library as Django 5.0 deprecates pytz.

ColeDCrawford committed:
Update Python PATHs in test Dockerfile

ColeDCrawford committed:
Fix field reference in signal receiver

prime_profile should be parent_profile

Improves error handling of the receiver handler. The old handler was simpler but made some assumptions about the existence and state of the prime_consumer attribute.

ColeDCrawford committed:
Add ZoneInfo backport for Python3.8 support

3.8 is EOL in October but easy enough to support it until then.

d-flood committed:
Merge pull request #6 from artshumrc/django-5

Django 5 fixes

ColeDCrawford committed:
Conditionally import ZoneInfo; use matrix strategy

Instead of using tox, use a build argument to control the Python Docker base image to take advantage of GHA parallel / matrix strategy. We end up with multiple containers but they should build faster and be easier to debug than having to manage multiple versions of Python and PATH in one image.

ColeDCrawford committed:
Update docs

ColeDCrawford committed:
Update package version

ColeDCrawford committed:
Merge pull request #7 from artshumrc/django-5

Conditionally import ZoneInfo; use GHA matrix strategy

ColeDCrawford committed:
Remove license key until we determine license

uv / hatch failing to build in uv MCH with an empty license (`ValueError: Error parsing field `project.license` - Invalid license expression: ''`)

ColeDCrawford committed:
Fix PyLD version syntax

d-flood committed:
remove url namespace

d-flood committed:
fix base.txt

d-flood committed:
Merge remote-tracking branch 'origin/django-5' into django-package

d-flood committed:
remove docker compose versions

d-flood committed:
update python versions matrix

d-flood committed:
update python versions in dockerfiles

d-flood committed:
include instructions for cutting a github release

d-flood committed:
update version to 3.0.0

ColeDCrawford committed:
Finish dropping Python 3.8 support

- Remove backports (unneeded for 3.9+)
- Update metadata in pyproject.toml

ColeDCrawford committed:
Remove backports from pyproject.toml dependencies too

ColeDCrawford committed:
Pin postgres:16 in primary compose config
